### PR TITLE
Added Reddit.restoreAuthenticatedInstance + version roll

### PR DESCRIPTION
**Pull request by bkonyi**
of a Reddit instance using previously cached credentials.

Also:
* Rolled version to 0.1.5
* Updated CHANGELOG.md
* Cleaned up some documentation.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,12 +1,18 @@
...
+* Various bug fixes.
+
 ## Version 0.1.4 (2018/03/31)
 * Added `SubredditModeration`, a class which implements moderator functionality for `Subreddit`s.
 * Fixed issue #46 which was causing `WebAuthenticator.url` to hit an assertion when `compactLogin` was set
 to `true`.
 
 ## Version 0.1.3 (2018/03/22)
-* Added additional convenience accessors to various classes, including `Comment`, `Redditor`, `Submission`, 
+* Added additional convenience accessors to various classes, including `Comment`, `Redditor`, `Submission`,
```
Comment (ckartik) [inline CHANGELOG.md]: I'm all for the removal of unnecessary whitespace, but just want to confirm this was an intentional edit.

File: lib/src/auth.dart [hunk 2]
```diff
@@ -335,12 +343,17 @@ class WebAuthenticator extends Authenticator {
...
   static WebAuthenticator create(
-      DRAWConfigContext config, oauth2.AuthorizationCodeGrant grant) {
-    final WebAuthenticator authenticator =
-        new WebAuthenticator._(config, grant);
-    return authenticator;
-  }
+          DRAWConfigContext config, oauth2.AuthorizationCodeGrant grant) =>
+      new WebAuthenticator._(config, grant);
+
+  static WebAuthenticator restore(
```
Comment (ckartik) [inline lib/src/auth.dart]: Why did we decide to wrap these public functions wrappers to private implementations?

File: README.md [hunk 3]
```diff
@@ -51,28 +51,87 @@ Here is a simple example of how to use web authentication with DRAW:
...
+    // Assuming the `code` query parameter is stored in a variable
+    // `auth_code`, we pass it to the `authorize` method in the
+    // `WebAuthenticator`.
+    await reddit.auth.authorize(auth_code);
+
+    // Write credentials to disk.
+    await writeCredentials(reddit.auth.credentials.toJson());
+  } else {
+    // Create a new Reddit instance using previously cached credentials.
+    reddit = await Reddit.restoreAuthenticatedInstance(
```
Comment (ckartik) [inline README.md]: So I read up on **Futures and Error Handling**, but I can't seem to find how we would handle a future that throws, when we have await.

So my question is:
1. Should we have a `try catch` here or does the _Future_ return `null`, when it resolves with an error during the use of `await`?

https://www.dartlang.org/guides/libraries/futures-error-handling

File: lib/src/reddit.dart [hunk 4]
```diff
@@ -125,6 +128,75 @@ class Reddit {
...
+      String clientSecret,
+      String userAgent,
+      String credentialsJson,
+      Uri redirectUri,
+      Uri tokenEndpoint,
+      Uri authEndpoint,
+      Uri configUri,
+      String siteName = 'default'}) async {
+    if (credentialsJson == null) {
+      throw new DRAWArgumentError('credentialsJson cannot be null.');
```
Comment (ckartik) [inline lib/src/reddit.dart]: Should we enforce the `credentialsJson` to be a required parameter in the function contract?
i.e
```dart
static Future<Reddit> restoreAuthenticatedInstance(String credentialsJson,
{String clientId,
       String clientSecret,
       String userAgent,
       String credentialsJson,
       Uri redirectUri,
       Uri tokenEndpoint,
       Uri authEndpoint,
       Uri configUri,
       String siteName = 'default'}) async {
```

File: CHANGELOG.md (hunk 1 above)
Comment (bkonyi) [reply CHANGELOG.md]: Yup, totally intentional!

File: lib/src/auth.dart (hunk 2 above)
Comment (bkonyi) [reply lib/src/auth.dart]: In this case, I'm actually not sure anymore haha.

File: README.md (hunk 3 above)
Comment (bkonyi) [reply README.md]: The `Future` isn't completed when `await` is used and an exception is thrown (or if it is, it doesn't matter anymore) which can be caught in a `catch`. I don't think we should need one here for just the examples.

File: lib/src/reddit.dart [hunk 5]
```diff
@@ -206,6 +289,7 @@ class Reddit {
   }
 
   RedditorRef redditor(String redditor) => new RedditorRef.name(this, redditor);
+
```
[the remark this answers is not in the document]
Comment (bkonyi) [reply lib/src/reddit.dart]: Don't think so... huh.

File: lib/src/reddit.dart (hunk 4 above)
Comment (bkonyi) [reply lib/src/reddit.dart]: Yeah, it might be better to do it your way. It's a shame that there's no good way to have required named parameters :(

File: lib/src/reddit.dart (hunk 5 above)
[the remark this answers is not in the document]
Comment (bkonyi) [reply lib/src/reddit.dart]: Yeah, that was intentional, was just hard to tell in the diff.
